test/extensions/http/cache/file_system_http_cache/file_system_http_cache_test.cc

```diff
@@ -116,12 +121,75 @@ TEST_F(FileSystemHttpCacheTestWithNoDefaultCache, InitialStatsAreSetCorrectly) {
   env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-b"), file_2_contents, true);
   cache_ = std::dynamic_pointer_cast<FileSystemHttpCache>(
       http_cache_factory_->getCache(cacheConfig(cfg), context_));
+  waitForEvictionThreadIdle();
   EXPECT_EQ(cache_->stats().size_limit_bytes_.value(), max_size);
   EXPECT_EQ(cache_->stats().size_limit_count_.value(), max_count);
   EXPECT_EQ(cache_->stats().size_bytes_.value(), file_1_contents.size() + file_2_contents.size());
   EXPECT_EQ(cache_->stats().size_count_.value(), 2);
 }
 
+TEST_F(FileSystemHttpCacheTestWithNoDefaultCache, EvictsOldestFilesUntilUnderCountLimit) {
+  const std::string file_contents = "XXXXX";
+  const uint64_t max_count = 2;
+  ConfigProto cfg = testConfig();
+  cfg.mutable_max_cache_entry_count()->set_value(max_count);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-a"), file_contents, true);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-b"), file_contents, true);
+  {
+    // TODO(ravenblack): replace this with backdating the files when that's possible.
+    absl::Mutex mu;
+    absl::MutexLock lock(&mu);
+    auto cond = []() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu) { return false; };
+    mu.AwaitWithTimeout(absl::Condition(&cond), absl::Seconds(1));
+  }
+  cache_ = std::dynamic_pointer_cast<FileSystemHttpCache>(
+      http_cache_factory_->getCache(cacheConfig(cfg), context_));
+  waitForEvictionThreadIdle();
+  EXPECT_EQ(cache_->stats().size_bytes_.value(), file_contents.size() * 2);
+  EXPECT_EQ(cache_->stats().size_count_.value(), 2);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-c"), file_contents, true);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-d"), file_contents, true);
+  cache_->trackFileAdded(file_contents.size());
+  cache_->trackFileAdded(file_contents.size());
+  waitForEvictionThreadIdle();
+  EXPECT_EQ(cache_->stats().size_bytes_.value(), file_contents.size() * 2);
+  EXPECT_EQ(cache_->stats().size_count_.value(), 2);
+  EXPECT_FALSE(Filesystem::fileSystemForTest().fileExists(absl::StrCat(cache_path_, "cache-a")));
+  EXPECT_FALSE(Filesystem::fileSystemForTest().fileExists(absl::StrCat(cache_path_, "cache-b")));
+  EXPECT_TRUE(Filesystem::fileSystemForTest().fileExists(absl::StrCat(cache_path_, "cache-c")));
+  EXPECT_TRUE(Filesystem::fileSystemForTest().fileExists(absl::StrCat(cache_path_, "cache-d")));
+}
+
+TEST_F(FileSystemHttpCacheTestWithNoDefaultCache, EvictsOldestFilesUntilUnderSizeLimit) {
+  const std::string file_contents = "XXXXX";
+  const std::string large_file_contents = "XXXXXXXXXXXX";
+  const uint64_t max_size = large_file_contents.size();
```

Comment:
I don't think it's worth it here, there isn't really that much common (except the sleep-hack), and abstracting small pieces away would make it harder to follow what the test is actually doing.

Reply:
OK. FWIW it might be worth scanning through the tests for mod_pagespeed's file cache: https://github.com/apache/incubator-pagespeed-mod/blob/master/test/pagespeed/kernel/cache/file_cache_test.cc

Some of these may not be relevant due to different eviction architecture (periodic cleaning rather than triggered by a write).